Containers with background color

duartecaldascardoso opened the pull request:
## Describe your changes
- Added a new optional parameter color to st.container. 
- Set the possible colors to be used in the color parameter to be the same as the ones in the markdown background
- Parameter is present and compatible in containers with height and border parameters 


## GitHub Issue Link
Closes #10531 

## Testing Plan

✅ Added Unit Tests (Python):

- In order to test the new color parameter from the container element, tests were added to make sure that there was flexibility in the backend to use any string as the input for the color parameter.

---

## Complication
As described in the issue, the background color from the container had to be applied to the container and to the elements inside it which are of type chart or transparent. I messed around with applying inheritance of the styling to the child containers but only obtained evasive solutions. The current workaround is to define the background of the charts as transparent in order for them to have the wanted background color (from the container). Example:
![image](https://github.com/user-attachments/assets/402d5378-866c-452b-85f5-8be2773ba650)


**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

duartecaldascardoso committed:
Initial iteration of the background color prop

duartecaldascardoso committed:
Merge branch 'develop' into enhancement-10531

duartecaldascardoso committed:
Fixing mistake in the background attribution

duartecaldascardoso committed:
Code review and improvements

duartecaldascardoso committed:
Updating the color origin

duartecaldascardoso committed:
Code review

duartecaldascardoso committed:
Adding tests for the backend solution

duartecaldascardoso committed:
Code review

duartecaldascardoso commented:
@jrieke Created this pull request for the container background color enhancement issue. In the complication section of the PR description I explain my thoughts on the transparent/chart elements which you said to be the tricky part of the implementation

jrieke commented:
Awesome, thank you! I'm on vacation this week but will already let the eng team have a look.

duartecaldascardoso committed:
Code review improvements

duartecaldascardoso committed:
Merge branch 'develop' into enhancement-10531

sfc-gh-lwilby commented on frontend/lib/src/components/core/Block/Block.tsx:
```diff
@@ -278,8 +278,15 @@ function ScrollToBottomVerticalBlockWrapper(
+
+  // If we get a color from the proto, we set the background color to the corresponding color from the existing theme.
+  const backgroundColor = color
+    ? getContainerBgColors(theme)[`${color}bg`]
```
I think it might be better to have a function `getContainerBgColor` that takes the theme and color string and returns the color. That way you can also add tests.

sfc-gh-lwilby commented on lib/streamlit/elements/layouts.py:
```diff
@@ -77,6 +78,11 @@ def container(
             border is shown if the container is set to a fixed height and not
             shown otherwise.
 
+        color: str or None
```
What do you think about adding more explicit typing here for which strings are valid background colours?

I was thinking we could define a Literal listing valid strings that correspond to background colours. 

We could also add some validation, so that if the user passes in an invalid string we give them an error or warning to help them with debugging their code.

duartecaldascardoso replied:
This makes sense. I went with the basic approach of defining a String and then using that background color method in the frontend - trying to replicate the markdown behavior. 
Your suggestions is more user friendly so Im happy to implement it.

sfc-gh-lwilby commented on proto/streamlit/proto/Block.proto:
```diff
@@ -40,6 +40,8 @@ message Block {
     // Height of the container, activates scrolling
     uint32 height = 2;
+    // Background color for the container
+    string color = 3;
```
```suggestion
    Enum Color {
      ....
    }
    Color background_color = 3;
```

lukasmasuch replied:
nit: from past experience, it's better to use more flexible types in proto if there is any chance that we might want to extend the support (e.g., allow hex values). Reason is our proto backwards compatibility, which would require keeping this field around as unused tech debt..

sfc-gh-lwilby replied:
Yeah, I am OK with leaving out the enum since that is consistent with what we are already doing and for the reason Lukas mentioned. 

Another way we could approach extensibility is by using a message like this:

```
Enum ColorName {
   ...
}
message ColorConfig {
  oneof background_color_spec {
    ColorName color_name = 1;
    // potential future
    string hex_value = 2;
  }
}
```

My thinking is just to leave less to interpreting string values in the frontend, which is a recommendation in the [API Best Practices for protobuf](https://protobuf.dev/best-practices/api/). But I was on the fence about this suggestion because this is admittedly less complex than the examples they are using in the best practices.

sfc-gh-lwilby commented on lib/streamlit/elements/layouts.py:
```diff
@@ -46,6 +46,7 @@ def container(
         *,
         height: int | None = None,
         border: bool | None = None,
+        color: str | None = None,
```
```suggestion
        background_color: str | None = None,
```

sfc-gh-lwilby commented:
It would be good to add some E2E playwright tests as well.

sfc-gh-lwilby commented:
Thanks for adding this @duartecaldascardoso! I just have a few suggestions.

duartecaldascardoso commented:
@sfc-gh-lwilby thanks for the suggestions, I will run by them in the next few days 

Just one thing. I got around to implement the Literal for the allowed strings in coloring - but am not sure if I should implement the Enum, since this would mean maintaining the color values in two places, since the validation is already made upon the Literal. What do you think?

duartecaldascardoso committed:
Code review according to merge comments

duartecaldascardoso committed:
Updated code according to review

duartecaldascardoso committed:
Added playwright tests for the container

duartecaldascardoso committed:
fixing wrong test logic

duartecaldascardoso commented:
@sfc-gh-lwilby Hey, I implemented the suggestions from your review apart from defining the Enum for the Proto. 
I am not sure about typing standards used in the errors, however I added a new error for the wrong color values and am invoking it upon the string not matching one of the literal options. 
I also added some playwright tests for the container, including one with a transparent chart.

duartecaldascardoso committed:
Code review

sfc-gh-lwilby commented on frontend/lib/src/components/core/Block/Block.tsx:
```diff
@@ -278,8 +278,13 @@ function ScrollToBottomVerticalBlockWrapper(
+  const color = props.node.deltaBlock.vertical?.backgroundColor || undefined
+
+  // If we get a color from the proto, we set the background color to the corresponding color from the existing theme.
+  const backgroundColor = color ? getContainerBgColor(theme, color) : undefined
```
nit: we could call this function inside `VerticalBlockBorderWrapper` where we already have the theme and just pass in the value from the proto.

sfc-gh-lwilby commented on frontend/lib/src/theme/getColors.ts:
```diff
@@ -183,6 +183,30 @@ export function getMarkdownBgColors(theme: EmotionTheme): any {
+
+  const colorMap: Record<string, string> = {
+    red: theme.colors[lightTheme ? "red80" : "red60"],
+    orange: theme.colors.yellow70,
```
It looks using the same color for orange and yellow here is based on the markdown implementation @duartecaldascardoso ?

@jrieke do you know why we don't use orange for the orange background, it looks like something is defined for orange in `colors.ts`.

sfc-gh-lwilby commented on e2e_playwright/st_container.py:
```diff
@@ -51,3 +54,20 @@
     for i in range(10):
         st.chat_message("user").write(f"Message {i}")
+
+with st.container(background_color="red", height=200):
```
Thanks for adding these. To completely test this, you need to add some tests in `st_container_test.py`. You could add snapshots. 

Once you add the snapshot tests, on the first CI run it will fail since the snapshots are missing and then locally you can run `make update-snapshots` to add the snapshots into the correct directory and commit them. We have this page with info [about the e2e tests](https://github.com/streamlit/streamlit/wiki/Running-e2e-tests-and-updating-snapshots).

sfc-gh-lwilby commented:
> @sfc-gh-lwilby thanks for the suggestions, I will run by them in the next few days
> 
> Just one thing. I got around to implement the Literal for the allowed strings in coloring - but am not sure if I should implement the Enum, since this would mean maintaining the color values in two places, since the validation is already made upon the Literal. What do you think?

I would argue they are already maintained in two places since `getContainerBgColor` also maintains a list of all the expected strings mapped to the theme. If we had the enum I would imagine using the enum values in `getContainerBgColor` instead of relying on matching string values, this maintains a bit more of a contract between the python code in `layouts.py` and the FE code since they are both using the enum values. 

I will reply more about this above since Lukas also weighed in.

sfc-gh-lwilby commented on frontend/lib/src/theme/getColors.ts:
```diff
@@ -183,6 +183,30 @@ export function getMarkdownBgColors(theme: EmotionTheme): any {
   }
 }
 
+export function getContainerBgColor(
```
Could you add some simple tests for this function as well including edge cases like undefined and unrecognized colors? I know we are validating the colors in the python code, but just out of caution in case something were to change.

duartecaldascardoso replied:
Hey @sfc-gh-lwilby what would be the convention here? Shall I create a tests class solely for this or should I append it to some utils testing class? I noticed there are no tests for the other coloring related methods

sfc-gh-lwilby replied:
We typically just create test files in the same directory with the same name as the file containing the functions to test. So, `getColors.test.ts` here.

sfc-gh-lwilby commented:
Thanks @duartecaldascardoso for making these updates, I just have a few more suggestions and test requests, and I want to confirm the orange color with Johannes also before merging. 

Thanks so much for working on this! People are going to be happy about this feature.

duartecaldascardoso commented:
@sfc-gh-lwilby thanks for your suggestions, I will work on them in the next few days

duartecaldascardoso committed:
Moving the background logic to the vertical block

duartecaldascardoso committed:
Merge branch 'develop' into enhancement-10531

duartecaldascardoso committed:
Adding tests for playwright

duartecaldascardoso committed:
Merge branch 'enhancement-10531' of https://github.com/joaooliveira-11/streamlit into enhancement-10531

sfc-gh-lwilby commented:
@duartecaldascardoso I ran your CI for you, it looks like there are some errors in the playwright tests that need to be fixed before reaching the snapshot code. 

I want to mention as well, that I have a [big refactor PR](https://github.com/streamlit/streamlit/pull/10960) I have been working on that will conflict with this PR. I might merge it today. It will mean that the new proto field you are adding will need to be moved because I've added a new `FlexContainer` type that will be used by containers going forward. 

I can make the changes to your PR or develop (depending on which merges first) if you would prefer, or would you rather make the changes after I merge this?

duartecaldascardoso commented:
@sfc-gh-lwilby Honestly I won't have much time in the next week to dedicate here. I organized my time to implement the suggestions you gave but not much more than that. If you are okay with doing that work, all good by me! If not, I can get to that asap

duartecaldascardoso committed:
Update lib/tests/streamlit/elements/layouts_test.py

Co-authored-by: graphite-app[bot] <96075541+graphite-app[bot]@users.noreply.github.com>

sfc-gh-lwilby commented:
> @sfc-gh-lwilby Honestly I won't have much time in the next week to dedicate here. I organized my time to implement the suggestions you gave but not much more than that. If you are okay with doing that work, all good by me! If not, I can get to that asap

Yes, no problem, I can make the changes. I just wanted to check with you since it is your PR.

jrieke commented:
Hey, I just tried this out and there are a few things we need to change:

- If a background is shown, the container should have a padding and rounded borders similar to if `border=True` is set. 

  Current:
  
  ![Screenshot 2025-05-21 at 18 20 03](https://github.com/user-attachments/assets/d65c51ae-221f-4197-8850-be57bda3d352)
  
  Desired (I set `border=True` here, so that should obviously not show up if the border isn't shown):
  ![Screenshot 2025-05-21 at 18 20 26](https://github.com/user-attachments/assets/5b55df33-dc88-4fba-bb39-6ef6a4e4fddc)

- Only the colors that are allowed for colored Markdown text should be supported, namely blue, green, orange, red, violet, gray/grey, or primary. (We can add other colors like yellow but we should do it everywhere and need to figure out the exact color first, so better to do in a separate PR). 

- Charts don't seem to have the correct background color for me. Altair and Plotly charts still have a white background. Note that we should also try this out with all other elements and chart types. 

![Screenshot 2025-05-21 at 18 23 51](https://github.com/user-attachments/assets/75be1833-4e28-493b-843f-d8d2bbe6e815)

- Buttons should probably also be adapted to the background color of the container. Currently they simply have the background color (white by default). 
  ![Screenshot 2025-05-21 at 18 25 32](https://github.com/user-attachments/assets/9ea1e0ac-81a9-4b67-9487-6b3b52eab7db)
  
  In contrast, if you change the background color of the app via theming, the button color adapts to that. We should use the same mechanism in containers with a background color. Ideally for all elements, button is just the first one that comes to mind. 
  ![Screenshot 2025-05-21 at 18 26 49](https://github.com/user-attachments/assets/bff5edf6-3bb6-46f5-a0e4-0381c57ae1b8)

I also still need to make up my mind about the parameter name before we merge this – I think `color`, `background`, and `background_color` are all viable options but I need to think through which other, similar parameters we might want on other elements in the future.

jrieke commented:
Wanted to add here that we also discussed this feature in our team meeting today. Some of the discussion points that came up (none with definitive answers yet, just want to list them here!):

- Should we allow using the `secondaryBackgroundColor` as a container background color? E.g., with `st.container(background_color="secondary_background")`. And the other colors could then be inverted in this container, similar to how we do it in the sidebar. This would allow creating a typical dashboard look where you have a gray background with white cards on top. 
- Widgets inside a container with a background color can look weird if the widget color doesn't complement the background color. Not sure how we should work around this – are we just OK with this and tell devs they need to take care of this themselves? Or should we always use inverted colors when a container has a background color, similar to the sidebar? Or do we only allow setting the container background to the secondary background color, and not to any other color? 
- A more involved change that could mitigate some of the issues above would be to not simply set a background color on the container, but a full theme, that would also allow setting a `secondaryBackgroundColor` and other colors. E.g. `st.container(theme="main"|"sidebar"|"foo")`. This would require us to introduce named themes and probably require a lot more implementation time.

jrieke commented:
Going to revert this back to a draft for now since there are still a bunch of changes needed :)

lukasmasuch commented:
We could also allow direct overwriting of specific theme properties as proposed here: https://github.com/streamlit/streamlit/issues/10749:

```python
st.container(theme={"background_color": "red"})
```

This would resolve a large number of styling requests in a manageable way. The important part to make it work is by overwriting it in the theme context of the container, thereby all the child components will correctly adapt to this. 

I think we can also make it work with just a `background` parameter if we do it via theme context. We could also use the secondary background automatically if `background=True`.

jrieke commented:
Going to close this PR for now because there's still a lot of stuff left to do and it doesn't seem like anybody still wants to work on this at the moment. @duartecaldascardoso if you or someone else wants to take this further, please feel free to re-open this PR or create a new one!

And @duartecaldascardoso would still love to send you some swag for your effort, please feel free to fill out [this form](https://docs.google.com/forms/d/e/1FAIpQLSe7cXh3H1DmrgNpVew9qViGIHX7vdEbTv5isA44_z5bgaKTKg/viewform) and we'll send it to you! :)

duartecaldascardoso commented:
@jrieke makes sense, thanks for clarifying. Will be on the lookout :) 

Alright, I will, thanks!
